### PR TITLE
[Hackathon] Typescript (Only types, no refactorings)

### PR DESCRIPTION
**Pull request by mschuwalow**
Short description if any.

Author Todo List:

- [x] Add/adjust tests (if applicable)
- [x] Build in CI passes
- [x] Latest master revision is merged into the branch
- [x] Self-Review
- [x] Set `Ready For Review` status

### CODE REVIEW

**Review by wi101 — APPROVED** [review]

File: src/utils/types.ts [hunk 1]
```diff
@@ -1,126 +1,84 @@
...
+  return isNonEmpty(value) ? ([param, isFunction(transform) ? transform(value) : value] as [string, string]) : []
 }
 
-export function asStringParam (param, value) {
+export function asStringParam (param: string, value: string | number | boolean): [string, string] | [] {
   return asParamOrEmpty(param, value, (s) => encodeURIComponent(s))
 }
 
-export function asStringParamTransform (param, value, transform) {
+export function asStringParamTransform <A> (param: string, value: A, transform: (A) => string): [string, string] | [] {
```
Comment (wi101) [inline src/utils/types.ts:64]:
```suggestion
export function asStringParamTransform <A> (param: string, value: A, transform: (a: A) => string): [string, string] | [] {
```

**Review by wi101 — COMMENTED** [review]
Some comments otherwise it looks good to me

File: src/utils/page.ts [hunk 2]
```diff
@@ -1,24 +1,15 @@
...
-export function loadedDomain () {
+export function loadedDomain (): string {
   return (document.domain || (document.location && document.location.host)) || (window && window.location && window.location.host) || 'localhost'
 }
 
-/**
- * @return {string|undefined}
- */
-export function getReferrer (win = window) {
+export function getReferrer (win = window): string | undefined {
```
Comment (leonelcuevas) [inline src/utils/page.ts:8]: Maybe missing a type annotation for `win`? Same below

File: src/types.ts [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+export interface IStateWrapper {
+    data: State,
+    sendsPixel (): boolean,
+    combineWith (newInfo: State): IStateWrapper,
+    asTuples (): [string, string][]
+    asQuery (): IQuery,
+  }
+
+export interface IPixelSender {
```
Comment (leonelcuevas) [inline src/types.ts:184]: Just curiosity: what's the meaning of the `I` prefixing a lot of the types?

**Review by leonelcuevas — APPROVED** [review]

File: src/pixel/stringify.ts [hunk 4]
```diff
@@ -3,6 +3,6 @@ import { trim } from '../utils/types'
 
 export const MASK = '*********'
 
-export function replacer (key, value) {
+export function replacer (key, value: string): string {
```
[the remark this answers is not in the document]
Comment (mschuwalow) [reply src/pixel/stringify.ts:6]: Default type without any annotation is any. Added it explicitly for clarity

File: src/utils/page.ts (hunk 2 above)
Comment (mschuwalow) [reply src/utils/page.ts:8]: updated 👍

**Pull request merged**